load_test.py

```diff
@@ -14,16 +14,16 @@ def run_main_process(concurrency, duration, dataset, dataset_q, stop_q):
     logging.info("Test from main process")
 
     # Initialize the dataset_queue with 4*concurrency requests
-    for query in dataset.get_next_n_queries(4 * concurrency):
+    for query in dataset.get_next_n_queries(2 * concurrency):
         dataset_q.put(query)
 
     start_time = time.time()
     current_time = start_time
     while (current_time - start_time) < duration:
         # Keep the dataset queue full for duration
```

Comment:
do the user threads stop when the queue is empty 🤔 🤔 🤔 
if so, there's a race with the `time.sleep(0.1)` below :/

Reply:
Yeah, I should probably remove the time.sleep. The original intention was to put more items in the queue at a time, and keep the main process less busy. But the queue isn't written to hold a long list of elements.

Reply:
can't you have a simpler mechanism, without a queue, only with a mutex in `dataset.get_next_n_queries(1)` and all the users calling it? 
(I'm not sure how llm-load-test works with threads&processes though, I didn't look at it)

Reply:
> do the user threads stop when the queue is empty

The user processes don't exit but they will pause and wait for a new element.

> can't you have a simpler mechanism, without a queue, only with a mutex in dataset.get_next_n_queries(1) and all the users calling it?

This sounds like it would probably be simpler. One idea with the queue was that later on we could configure the rate at which the main process adds items to the queue, as a rate limit

Reply:
> The user processes don't exit but they will pause and wait for a new element.

ack, I'll have to look at the code sooner than later to better understand :)

> One idea with the queue was that later on we could configure the rate at which the main process adds items to the queue, as a rate limit

this would limit the rate a the test level. Another way (compatible with the mutex) to do it would be at the user level. Would that be acceptable ... or not ?